File: feature/compiler/src/main/java/org/robok/engine/feature/compiler/java/JavaCompiler.java
package org.robok.engine.feature.compiler.java;

/*
 *  This file is part of Robok © 2024.
 *
 *  Robok is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Robok is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with Robok.  If not, see <https://www.gnu.org/licenses/>.
 */

import android.content.Context;
import com.android.tools.r8.D8;
import com.android.tools.r8.D8Command;
import dalvik.system.DexClassLoader;
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.OutputStream;
import java.io.PrintStream;
import java.lang.reflect.InvocationTargetException;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import org.eclipse.jdt.internal.compiler.batch.Main;
import org.robok.engine.feature.compiler.Decompress;

/*
 * Class that compiles a java code, based on an File Java
 * @author Aquiles Trindade (trindadedev13).
 */
public final class JavaCompiler {

  private final Context context;
  private final List<String> logs = new ArrayList<>();

  public JavaCompiler(Context context) {
    this.context = context;
  }

  /*
   * Compile the code using ECJ.
   * @param compileItem a Item with Info to compile
   */
  public final void compile(final CompileItem compileItem) {
    logs.clear();
    if (!compileItem.getJavaFile().exists()
        || !compileItem.getJavaFile().getName().endsWith(".java")) {
      newLog("Invalid file: " + compileItem.getJavaFile().getAbsolutePath());
      return;
    }

    var errorWriter = new StringWriter();
    var errWriter = new PrintWriter(errorWriter);

    var outputWriter = new StringWriter();
    var outWriter = new PrintWriter(outputWriter);

    var outputDir = new File(compileItem.getOutputDir(), "classes/");
    if (!outputDir.exists() && !outputDir.mkdirs()) {
      newLog("failed create output directory: " + outputDir.getAbsolutePath());
      return;
    }

    var args = new ArrayList<String>();
    args.add("-1.8");
    args.add("-proc:none");
    args.add("-nowarn");
    args.add("-d");
    args.add(outputDir.getAbsolutePath());
    args.add("-sourcepath");
    args.add(" ");
    args.add(compileItem.getJavaFile().getAbsolutePath());
    args.add("-cp");
    args.add(getLibs());

    var compiler = new Main(outWriter, errWriter, false, null, null);
    var success = compiler.compile(args.toArray(new String[0]));

    if (!success) {
      newLog("failed to compile:\n" + errorWriter);
      return;
    }

    newLog("compiled with successful:\n" + outputWriter);

    try {
      var inputPath = outputDir.getAbsolutePath();
      var outputPath = outputDir.getAbsolutePath() + "/classes.jar";
      var jarPackager = new JarCreator(inputPath, outputPath);
      jarPackager.create();
      runOldD8(outputDir);
    } catch (IOException e) {
      newLog(e.toString());
    }
  }

  private final void runOldD8(final File outputDir) {
    try {
      var d8Args = new ArrayList<String>();
      d8Args.add("--release");
      d8Args.add("--min-api");
      d8Args.add("21");
      d8Args.add("--lib");
      d8Args.add(getAndroidJarFile().getAbsolutePath());
      d8Args.add("--output");
      d8Args.add(outputDir.getAbsolutePath());
      var classes = getClassFiles(new File(outputDir.getAbsolutePath()));
      for (var file : classes) {
        if (file.getName().endsWith(".class")) {
          d8Args.add(file.getAbsolutePath());
        }
      }

      D8.main(d8Args.toArray(new String[0]));
      run(outputDir);
    } catch (Exception e) {
      newLog(e.toString());
    }
  }

  /*
   * Run the compiled code with R8 & DexClassLoader
   * @param outputDir The path where classes.jar is located
   */
  public final void run(final File outputDir) { 
    try {
      var resultStr = new StringBuilder();
      var outputStream =
        new OutputStream() {
          @Override
          public void write(int v) {
            resultStr.append(String.valueOf((char) v));
          }
          
          @Override
          public String toString() {
            return resultStr.toString();
          }
        };
        
      System.setOut(new PrintStream(outputStream));
      System.setErr(new PrintStream(outputStream));
      
      var className = "Main";
      var optimizedDir = context.getDir("odex", Context.MODE_PRIVATE).getAbsolutePath();

      var dexLoader =
          new DexClassLoader(
              outputDir.getAbsolutePath() + "/classes.dex",
              optimizedDir,
              null,
              context.getClassLoader());

      Class<?> calledClass = dexLoader.loadClass(className);
      var method = calledClass.getDeclaredMethod("main", String[].class);
      String[] param = {};
      var result = method.invoke(null, new Object[] {param});
      newLog(resultStr.toString());
    } catch (ClassNotFoundException
        | NoSuchMethodException
        | IllegalAccessException
        | InvocationTargetException e) {
      newLog(e.toString());
    }
  }

  private final List<File> getClassFiles(final File dir) {
    var files = new ArrayList<File>();
    File[] fileArr = dir.listFiles();
    if (fileArr == null) {
      return files;
    }

    for (var file : fileArr) {
      if (file.isDirectory()) {
        files.addAll(getClassFiles(file));
      } else {
        files.add(file);
      }
    }

    return files;
  }

  public final void newLog(final String log) {
    logs.add(log);
  }

  public final List<String> getLogs() {
    return logs;
  }

  private final String getLibs() {
    var libs = new StringBuilder();
    libs.append(getAndroidJarFile().getAbsolutePath());
    libs.append(":");
    libs.append(getLambdaFactoryFile().getAbsolutePath());
    return libs.toString();
  }

  private final File getAndroidJarFile() {
    var androidJar = new File(context.getFilesDir() + "/temp/android.jar");

    if (androidJar.exists()) return androidJar;

    Decompress.unzipFromAssets(
        context, "android.jar.zip", androidJar.getParentFile().getAbsolutePath());

    return androidJar;
  }

  private final File getLambdaFactoryFile() {
    var lambdaFactory = new File(context.getFilesDir() + "/temp/core-lambda-stubs.jar");

    if (lambdaFactory.exists()) return lambdaFactory;

    Decompress.unzipFromAssets(
        context, "core-lambda-stubs.zip", lambdaFactory.getParentFile().getAbsolutePath());

    return lambdaFactory;
  }

  public static final class CompileItem {
    private final File javaFile;
    private final File outputDir;

    public CompileItem(final File javaFile, final File outputDir) {
      this.javaFile = javaFile;
      this.outputDir = outputDir;
    }

    public final File getJavaFile() {
      return javaFile;
    }

    public final File getOutputDir() {
      return outputDir;
    }
  }
}

File: feature/compiler/src/main/java/org/robok/engine/feature/compiler/java/JavaCompiler.kt
package org.robok.engine.feature.compiler.java

/*
 * Copyright 2025 Robok.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.content.Context
import com.android.tools.r8.D8
import dalvik.system.DexClassLoader
import java.io.*
import org.eclipse.jdt.internal.compiler.batch.Main
import org.robok.engine.feature.compiler.Decompress

/**
 * Class that compiles Java code from a Java file.
 *
 * @author Aquiles Trindade (trindadedev13)
 */
class JavaCompiler(private val context: Context) {

  private val logs = mutableListOf<String>()

  /**
   * Compiles the code using ECJ.
   *
   * @param compileItem An item with information to compile.
   */
  fun compile(compileItem: CompileItem) {
    logs.clear()
    if (!compileItem.javaFile.exists() || !compileItem.javaFile.name.endsWith(".java")) {
      newLog("Invalid file: ${compileItem.javaFile.absolutePath}")
      return
    }

    val errorWriter = StringWriter()
    val errWriter = PrintWriter(errorWriter)

    val outputWriter = StringWriter()
    val outWriter = PrintWriter(outputWriter)

    val outputDir = File(compileItem.outputDir, "classes/")
    if (!outputDir.exists() && !outputDir.mkdirs()) {
      newLog("Failed to create output directory: ${outputDir.absolutePath}")
      return
    }

    val args =
      mutableListOf<String>(
        "-1.8",
        "-proc:none",
        "-nowarn",
        "-d",
        outputDir.absolutePath,
        "-sourcepath",
        " ",
        compileItem.javaFile.absolutePath,
        "-cp",
        getLibs(),
      )

    val compiler = Main(outWriter, errWriter, false, null, null)
    val success = compiler.compile(args.toTypedArray())

    if (!success) {
      newLog("Failed to compile:\n$errorWriter")
      return
    }

    newLog("Compiled successfully:\n$outputWriter")

    try {
      val inputPath = outputDir.absolutePath
      val outputPath = "$inputPath/classes.jar"
      val jarPackager = JarCreator(inputPath, outputPath)
      jarPackager.create()
      runOldD8(outputDir)
    } catch (e: IOException) {
      newLog(e.toString())
    }
  }

  private fun runOldD8(outputDir: File) {
    try {
      val d8Args =
        mutableListOf(
          "--release",
          "--min-api",
          "21",
          "--lib",
          getAndroidJarFile().absolutePath,
          "--output",
          outputDir.absolutePath,
        )
      val classes = getClassFiles(outputDir)
      for (file in classes) {
        if (file.name.endsWith(".class")) {
          d8Args.add(file.absolutePath)
        }
      }

      D8.main(d8Args.toTypedArray())
      run(outputDir)
    } catch (e: Exception) {
      newLog(e.toString())
    }
  }

  /**
   * Runs the compiled code using R8 and DexClassLoader.
   *
   * @param outputDir The path where classes.jar is located.
   */
  fun run(outputDir: File) {
    try {
      val resultStr = StringBuilder()
      val outputStream =
        object : OutputStream() {
          override fun write(v: Int) {
            resultStr.append(v.toChar())
          }

          override fun toString(): String = resultStr.toString()
        }

      System.setOut(PrintStream(outputStream))
      System.setErr(PrintStream(outputStream))

      val className = "Main"
      val optimizedDir = context.getDir("odex", Context.MODE_PRIVATE).absolutePath

      val dexLoader =
        DexClassLoader(
          "${outputDir.absolutePath}/classes.dex",
          optimizedDir,
          null,
          context.classLoader,
        )

      val calledClass = dexLoader.loadClass(className)
      val method = calledClass.getDeclaredMethod("main", Array<String>::class.java)
      method.invoke(null, arrayOf<String>())
      newLog(resultStr.toString())
    } catch (e: Exception) {
      newLog(e.toString())
    }
  }

  private fun getClassFiles(dir: File): List<File> {
    val files = mutableListOf<File>()
    val fileArr = dir.listFiles() ?: return files

    for (file in fileArr) {
      if (file.isDirectory) {
        files.addAll(getClassFiles(file))
      } else {
        files.add(file)
      }
    }

    return files
  }

  private fun newLog(log: String) {
    logs.add(log)
  }

  fun getLogs(): List<String> = logs

  private fun getLibs(): String {
    return "${getAndroidJarFile().absolutePath}:${getLambdaFactoryFile().absolutePath}"
  }

  private fun getAndroidJarFile(): File {
    val androidJar = File("${context.filesDir}/temp/android.jar")

    if (androidJar.exists()) return androidJar

    Decompress.unzipFromAssets(context, "android.jar.zip", androidJar.parentFile!!.absolutePath)

    return androidJar
  }

  private fun getLambdaFactoryFile(): File {
    val lambdaFactory = File("${context.filesDir}/temp/core-lambda-stubs.jar")

    if (lambdaFactory.exists()) return lambdaFactory

    Decompress.unzipFromAssets(
      context,
      "core-lambda-stubs.zip",
      lambdaFactory.parentFile!!.absolutePath,
    )

    return lambdaFactory
  }

  data class CompileItem(val javaFile: File, val outputDir: File)
}
